### PR TITLE
Add setting to make "schedule based" items "event based"

### LINKED ISSUE #2014: Setting to make "schedule based" items "event based"
### Is your feature request related to a problem? Please describe.
Several users asked if it might be possible to have the DND mode as event based item, not schedule based. This makes totally sense to me; for wifi name as well. On Android 8 or higher listening to these event isn't possible anymore, when the app is in the background.

### Describe the solution you'd like
Add a setting that enables a foreground service running a broadcast receiver. Currently I've set the schedule to 15 minutes on my device and the battery usage of openHAB is about 1%. So the battery usage might be negligible.

### Describe alternatives you've considered
Having the broadcast receiver at least running when the app is in the foreground.

### Additional context

- https://community.openhab.org/t/android-app-beta/41382/236
- https://github.com/openhab/openhab-android/issues/1663#issuecomment-629592558

### PR DESCRIPTION
**Pull request by mueller-ma**
Closes #2014

Signed-off-by: mueller-ma <mueller-ma@users.noreply.github.com>

### CODE REVIEW

File: mobile/src/main/java/org/openhab/habdroid/background/ForegroundBroadcastReceiver.kt [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+import android.util.Log
+import androidx.annotation.VisibleForTesting
+import androidx.core.app.NotificationCompat
+import androidx.core.content.ContextCompat
+import org.openhab.habdroid.R
+import org.openhab.habdroid.util.PrefKeys
+import org.openhab.habdroid.util.getPrefs
+import org.openhab.habdroid.util.isItemUpdatePrefEnabled
+
+class ForegroundBroadcastReceiver : Service() {
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/background/ForegroundBroadcastReceiver.kt:30]: This class seems a bit misnamed. How about `BroadcastEventListenerService`?

File: mobile/src/main/java/org/openhab/habdroid/background/ForegroundBroadcastReceiver.kt [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    private var isServiceStarted = false
+    private var backgroundTasksManager = BackgroundTasksManager()
+    private var isRegistered = false
+
+    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
+        Log.d(TAG, "onStartCommand())")
+        when (intent?.action) {
+            ACTION_START -> startService()
+            ACTION_STOP -> stopService()
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/background/ForegroundBroadcastReceiver.kt:39]: This logic seems to be unneeded. I'd instead suggest to
- Call `context.stopService()` in `startOrStopService`
- Drop the actions completely
- Check whether there are matching intent filter actions unconditionally in onStartCommand() and call `stopSelf(startId)` if there aren't any
- Do receiver unregistering (if needed) in `onDestroy`

File: mobile/src/main/java/org/openhab/habdroid/background/ForegroundBroadcastReceiver.kt [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+                getString(R.string.send_device_info_foreground_service_running_summary_one, titlesOfItems[0])
+            }
+            titlesOfItems.size == 2 -> {
+                getString(R.string.send_device_info_foreground_service_running_summary_two,
+                    titlesOfItems[0], titlesOfItems[1])
+            }
+            else -> {
+                getString(R.string.send_device_info_foreground_service_running_summary_more,
+                    titlesOfItems[0], titlesOfItems[1])
+            }
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/background/ForegroundBroadcastReceiver.kt:69]: Wouldn't a summary like `Listening for 3 event types` be sufficient?

File: mobile/src/main/java/org/openhab/habdroid/background/ForegroundBroadcastReceiver.kt [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+import androidx.annotation.VisibleForTesting
+import androidx.core.app.NotificationCompat
+import androidx.core.content.ContextCompat
+import org.openhab.habdroid.R
+import org.openhab.habdroid.util.PrefKeys
+import org.openhab.habdroid.util.getPrefs
+import org.openhab.habdroid.util.isItemUpdatePrefEnabled
+
+class ForegroundBroadcastReceiver : Service() {
+    private var isServiceStarted = false
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/background/ForegroundBroadcastReceiver.kt:31]: This variable is unused.

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt [hunk 5]
```diff
@@ -283,7 +286,11 @@ class MainActivity : AbstractBaseActivity(), ConnectionFactory.UpdateListener {
         updateTitle()
         showMissingPermissionsWarningIfNeeded()
 
-        registerReceiver(backgroundTasksManager, BackgroundTasksManager.getIntentFilterForForeground())
+        val intentFilter = BackgroundTasksManager.getIntentFilterForForeground(this)
+        if (intentFilter.countActions() != 0 &&
+            !prefs.getBoolean(PrefKeys.SEND_DEVICE_INFO_FOREGROUND_SERVICE, false)) {
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt:291]: Add an extension func for that getter?

Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt]: Done

File: mobile/src/main/java/org/openhab/habdroid/background/ForegroundBroadcastReceiver.kt (hunk 4 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/background/ForegroundBroadcastReceiver.kt]: Done

File: mobile/src/main/java/org/openhab/habdroid/background/ForegroundBroadcastReceiver.kt (hunk 2 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/background/ForegroundBroadcastReceiver.kt]: Thanks, that looks way better.

File: mobile/src/main/java/org/openhab/habdroid/background/ForegroundBroadcastReceiver.kt (hunk 1 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/background/ForegroundBroadcastReceiver.kt]: Done

File: mobile/src/main/java/org/openhab/habdroid/background/ForegroundBroadcastReceiver.kt (hunk 3 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/background/ForegroundBroadcastReceiver.kt]: I'd like to inform the user why the event listener is running. Mentioning just the count is sufficient, too, but I like my version more.

**Comment by maniac103** [comment]
LGTM now from looking at the code ... I haven't actually tested it though.

**Pull request merged**